arrow-libs/suspendapp/suspendapp/src/commonMain/kotlin/arrow/continuations/SuspendApp.kt

```diff
@@ -17,38 +17,55 @@ import kotlinx.coroutines.*
  * @param timeout the maximum backpressure time that can be applied to the process. This emulates a
  *   `SIGKILL` command, and after the [timeout] is passed the App will forcefully shut down
  *   regardless of finalizers.
+ * @param forceExit whether to force-exit the process after this SuspendApp
+ *   finishes (code 0 on success or 255 on failure)
  * @param block the lambda of the actual application.
  */
-@OptIn(ExperimentalStdlibApi::class)
 public fun SuspendApp(
   context: CoroutineContext = Dispatchers.Default,
   uncaught: (Throwable) -> Unit = Throwable::printStackTrace,
   timeout: Duration = Duration.INFINITE,
   process: Process = process(),
+  forceExit: Boolean = true,
   block: suspend CoroutineScope.() -> Unit,
 ): Unit =
   process.use { env ->
-    env.runScope(context) {
-      val job =
-        launch(start = CoroutineStart.LAZY) {
-          try {
-            block()
-            env.exit(0)
-          } catch (_: SuspendAppShutdown) {} catch (e: Throwable) {
-            uncaught(e)
-            env.exit(-1)
+    val jobCause = CompletableDeferred<Throwable?>()
+    try {
+      env.runScope(context) {
+        val job =
+          launch(start = CoroutineStart.LAZY) {
+            try {
+              block()
+            } catch (_: SuspendAppShutdown) {} catch (e: Throwable) {
+              throw e
+            }
           }
-        }
-      val unregister =
-        env.onShutdown {
-          withTimeout(timeout) {
-            job.cancel(SuspendAppShutdown)
-            job.join()
+        val unregister =
+          env.onShutdown {
+            withTimeout(timeout) {
+              job.cancel(SuspendAppShutdown)
+              job.join()
+            }
           }
+        job.invokeOnCompletion {
+          unregister()
+          jobCause.complete(it)
+        }
+        job.start()
+      }
+    } catch(e: Throwable) {
+        uncaught(e)
+        throw e
```

Comment:
this throw will propagate only if `forceExit` is false - as otherwise the process should be shutdown before it will go anywhere